File: simple-pos-management-app/client/src/service/category.ts
 import api from './api';
import { Category, CategoryCreate, CategoryWithProducts } from '../types/category';

export async function getCategories() {
  const response = await api.get<Category[]>('/categories/');
  return response.data;
}

export async function createCategory(category: CategoryCreate) {
  const response = await api.post<Category>('/categories/', category);
  return response.data;
}

export async function getCategory(categoryId: number) {
  const response = await api.get<CategoryWithProducts>(`/categories/${categoryId}`);
  return response.data;
}

export async function updateCategory(categoryId: number, category: CategoryCreate) {
  const response = await api.put<Category>(
    `/categories/${categoryId}`,
    category
  );
  return response.data;
}

export async function deleteCategory(categoryId: number) {
  const response = await api.delete(`/categories/${categoryId}`);
  return response.data;
}

File: simple-pos-management-app/server/app/api/v1/endpoints/category.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
from app.schemas.category import Category, CategoryCreate, CategoryWithProducts
from app.db.session import get_db
from app.services import category_service

router = APIRouter()


@router.get("/categories/", response_model=List[Category])
async def get_categories(
    db: Session = Depends(get_db),
):
    return category_service.get_categories(db)


@router.post("/categories/", response_model=Category)
async def create_category(
    category: CategoryCreate,
    db: Session = Depends(get_db),
):
    existing_category = category_service.get_category_by_name(db, category.name)
    if existing_category:
        raise HTTPException(
            status_code=400, detail="Category with this name already exists"
        )
    return category_service.create_category(db, category)


@router.get("/categories/{category_id}", response_model=CategoryWithProducts)
async def get_category(
    category_id: int,
    db: Session = Depends(get_db),
):
    category = category_service.get_category_by_id(db, category_id)
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    return category


@router.put("/categories/{category_id}", response_model=Category)
async def update_category(
    category_id: int,
    category: CategoryCreate,
    db: Session = Depends(get_db),
):
    updated_category = category_service.update_category(db, category_id, category)
    if not updated_category:
        raise HTTPException(status_code=404, detail="Category not found")
    return updated_category


@router.delete("/categories/{category_id}")
async def delete_category(
    category_id: int,
    db: Session = Depends(get_db),
):
    result = category_service.delete_category(db, category_id)
    if result is False:
        raise HTTPException(status_code=404, detail="Category not found")
    if result == "has_products":
        raise HTTPException(
            status_code=400, detail="Cannot delete category with existing products"
        )
    return {"message": "Category deleted successfully"}
